libexec/python/docker/tasks.py

```diff
@@ -167,16 +167,34 @@ def extract_env(manifest):
     '''extract the environment from the manifest, or return None.
     Used by functions env_extract_image, and env_extract_tar
     '''
+    from defaults import DOCKER_PATH
+
     environ = get_config(manifest, 'Env')
     if environ is not None:
         if not isinstance(environ, list):
             environ = [environ]
 
+        # Default PATH components to subtract Docker set PATH
+
+        DOCKER_PATH = DOCKER_PATH.split(':')
+
         lines = []
         for line in environ:
             line = re.findall("(?P<var_name>.+?)=(?P<var_value>.+)", line)
-            line = ['export %s="%s"' % (x[0], x[1]) for x in line]
-            lines = lines + line
+
+            for x in line:
+                varname = x[0]
+                varval = x[1]
+
+                # For PATH, subtract Docker set PATH from Singularity default
+
+                if varname == "PATH":
+                    parts = varval.split(':')
+                    parts = [x for x in parts if x not in DOCKER_PATH]
+                    varval = "$PATH:%s" % ":".join(parts)
```

Comment:
This has the potential to re-arrange the order of the `PATH`.  (See docker://godlovedc/lolcow for example.) But I don't see a good way around it.

Reply:
I, for one, don't expect SINGULARITYENV_PATH to override the PATH inside the container. The previously working approach with prepending it is like what environment modules usually do and is intuitive.

If there is no access to SINGULARITYENV_PATH inside the container at runtime then what is the following?

DEBUG   [U=0,P=24127]      singularity_runtime_environment()         Transposing environment
DEBUG   [U=0,P=24127]      singularity_runtime_environment()         Converting envar 'SINGULARITYENV_HISTFILE' to 'HISTFILE' = '(null)'
DEBUG   [U=0,P=24127]      envar_set()                               Unsetting environment variable: HISTFILE
DEBUG   [U=0,P=24127]      singularity_runtime_environment()         Converting envar 'SINGULARITYENV_PATH' to 'PATH' = '/opt/blupf90:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin'
DEBUG   [U=0,P=24127]      envar_set()                               Setting environment variable: 'PATH' = '/opt/blupf90:/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin'

The problem is that the above PATH seems to be discarded right after that and the 10-docker.sh PATH is used instead.

Reply:
@moskalenko this is probably happening elsewhere (not in the Python). The code above is written at the time of build, so you don't have access to `SINGULARITYENV_PATH` at that time.